### PR TITLE
[Issue 64] fix bug: type assert cause panic

### LINKED ISSUE #64: producer pending queue peek last with empty queue cause panic
#### Expected behavior

[pending queue peek last operator](https://github.com/apache/pulsar-client-go/blob/11755b26da59f58c2675708af1470fe493d92e37/pulsar/impl_partition_producer.go#L296) should deal with return value **nil**

#### Actual behavior

case panic:

> panic: interface conversion: interface {} is nil, not *pulsar.pendingItem

#### Steps to reproduce

this is code error

#### System configuration
**Pulsar version**: master branch

### PR DESCRIPTION
**Pull request by mileschao**
Signed-off-by: miles.j.chao [miles.j.chao@gmail.com](miles.j.chao@gmail.com)

Fixes #64 

### Motivation

1. type assert may cause panic when ```PeekLast()``` return nil

### CODE REVIEW

File: pulsar/impl_partition_producer.go [hunk 1]
```diff
@@ -293,14 +293,17 @@ func (p *partitionProducer) internalFlushCurrentBatch() {
 func (p *partitionProducer) internalFlush(fr *flushRequest) {
 	p.internalFlushCurrentBatch()
 
-	pi := p.pendingQueue.PeekLast().(*pendingItem)
```
Comment (xujianhai666) [inline pulsar/impl_partition_producer.go]: In point of my view, if type is specific, meaning the item from pendingQueue is *pendingItem, we can not type assert with ok.

Comment (mileschao) [reply pulsar/impl_partition_producer.go]: > In point of my view, if type is specific, meaning the item from pendingQueue is *pendingItem, we can not type assert with ok.

As [comment on PeekLast](https://github.com/apache/pulsar-client-go/blob/7a9aa2b36371437b1c409cb21cb2056ded7f198d/util/blocking_queue.go#L40)

> // PeekLast return last item in queue without dequeing, return nil if queue is empty

PeekLast() returns nil if queue is empty.

panic happens when ```PeekLast``` returns nil

File: go.mod [hunk 2]
```diff
@@ -4,21 +4,16 @@ go 1.12
...
-	github.com/beefsack/go-rate v0.0.0-20180408011153-efa7637bb9b6
-	github.com/bmizerany/perks v0.0.0-20141205001514-d9a9656a3a4b
-	github.com/cespare/xxhash v1.1.0 // indirect
+	github.com/cespare/xxhash/v2 v2.1.0 // indirect
 	github.com/deckarep/golang-set v1.7.1
 	github.com/golang/protobuf v1.3.1
 	github.com/google/go-cmp v0.3.0 // indirect
-	github.com/inconshreveable/mousetrap v1.0.0 // indirect
-	github.com/klauspost/compress v1.6.2
+	github.com/klauspost/compress v1.8.4
```
Comment (wolfstudy) [inline go.mod]: Why we need to fix `go.mod`

**Comment by wolfstudy** [comment]
ping @mileschao PTAL

File: pulsar/impl_partition_producer.go [hunk 3]
```diff
@@ -293,14 +293,17 @@ func (p *partitionProducer) internalFlushCurrentBatch() {
...
+	if ok {
+		pi.sendRequests = append(pi.sendRequests, &sendRequest{
+			msg: nil,
+			callback: func(id MessageID, message *ProducerMessage, e error) {
+				fr.err = e
+				fr.waitGroup.Done()
+			},
+		})
+	}
+
```
Comment (wolfstudy) [inline pulsar/impl_partition_producer.go]: If not ok, `fr.waitGroup.Done()` will not be executed, `wg.wait()` in send will not be released, and sending will be blocked indefinitely

**Comment by wolfstudy** [comment]
ping @mileschao PTAL

**Pull request closed without merging**